File: THCS/2022_2023/LongAn_2223/Bai3/Bai3.cpp
#include <iostream>
#include <unordered_set>

using namespace std;

// Hàm tính tổng bình phương các chữ số của n
long long sum_of_squares_of_digits(long long n) {
    long long total = 0;
    while (n > 0) {
        int digit = n % 10;
        total += digit * digit;
        n /= 10;
    }
    return total;
}

// Hàm kiểm tra xem n có dẫn đến 1 không
bool leads_to_one(long long n) {
    unordered_set<long long> seen;  // Tập hợp để lưu các số đã thấy
    while (n != 1 && seen.find(n) == seen.end()) {
        seen.insert(n);
        n = sum_of_squares_of_digits(n);
    }
    return n == 1;
}

// Hàm tìm số K
long long find_k(long long n) {
    long long k = n + 1;  // Bắt đầu từ N + 1
    while (true) {
        if (leads_to_one(k)) {
            return k;
        }
        k++;  // Kiểm tra số tiếp theo
    }
}

int main() {
    long long N;
    cout << "Nhập số nguyên dương N: ";
    cin >> N;  // Nhập N từ bàn phím
    long long K = find_k(N);
    cout << K << endl;  // Xuất K ra màn hình
    return 0;
}

File: THCS/2022_2023/LongAn_2223/Bai3/Bai3.py
def sum_of_squares_of_digits(n):
    """Tính tổng bình phương các chữ số của n."""
    total = 0
    while n > 0:
        digit = n % 10
        total += digit ** 2
        n //= 10
    return total

def leads_to_one(n):
    """Kiểm tra xem n có dẫn đến 1 không."""
    seen = set()  # Tập hợp để lưu các số đã thấy
    while n != 1 and n not in seen:
        seen.add(n)
        n = sum_of_squares_of_digits(n)
    return n == 1

def find_k(n):
    """Tìm số K nhỏ nhất lớn hơn N mà dẫn đến 1."""
    k = n + 1  # Bắt đầu từ N + 1
    while True:
        if leads_to_one(k):
            return k
        k += 1  # Kiểm tra số tiếp theo

if __name__ == "__main__":
    N = int(input("Nhập số nguyên dương N: "))  # Nhập N từ bàn phím
    K = find_k(N)
    print(K)
